test/metabase/query_processor_test/date_time_zone_functions_test.clj

```diff
@@ -169,3 +177,141 @@
                  :fields [[:field (mt/id :times :index) nil]]}]]]
         (testing title
           (is (= expected (test-date-extract query))))))))
+
+(defn- date-math
+  [op x amount unit col-type]
+  (let [amount (if (= op :date-add)
+                 amount
+                 (- amount))
+        fmt    (cond
+                 ;; the :date column of :presto should have this format too,
+                 ;; but the test data we created for presto is datetime even if we define it as date
+                 (and (= driver/*driver* :presto) (#{:text-as-date} col-type))
+                 "yyyy-MM-dd"
+
+                 :else
+                 "yyyy-MM-dd HH:mm:ss")]
+    (t/format fmt (u.date/add x unit amount))))
+
+(defn- mongo-major-version [db]
+  (-> (get-in db [:details :version])
+      (str/split #"\.")
+      first
+      parse-long))
+
+(deftest date-math-tests
+  (mt/dataset times-mixed
+    (mt/test-drivers (disj (mt/normal-drivers-with-feature :date-arithmetics) :mongo)
+      (testing "date arithmetic with datetime columns"
+        (doseq [[col-type field-id] [[:datetime (mt/id :times :dt)] [:text-as-datetime (mt/id :times :as_dt)]]
```

Comment:
Suggestions:

Could you make column type part of the `doseq` as well? That way you have one mega `doseq` instead of three

These sorts of big `doseq`-based tests get sort of hard to read I think. Since a lot of the stuff below is declarative I'd consider moving what you can out into a top-level `def` and call it `date-math-test-cases` or something. Then the actual business logic can go in the test itself. It makes things a little easier on the eyes I think.

Just remember that you shouldn't be loading the dataset or calling `mt/id` in a top-level form... so your `def` should have column name (`:dt`) and the test itself can use that to get the actual Field

Reply:
For the query you could change it to be a function maybe e.g. `:query-fn (fn [field-id] ...)`. IDK. Or just have it look at a dynamic var. Doesn't really matter. I just want this test to be easier to read

Reply:
If we want to move this to a top-level def, then the expected need to be an `fn` too, because the `expected` are being generated dynamically.

So it'll probably look something like

```clojure
{:expected-fn (fn [col-type op unit col-type]... )
 :query-fn    (fn [field-id] ...))
```
I'm not sure this makes it easier to reason about the tests tho.

But I'm all ears on how to improve these tests, I've been going back and forth with it several times and still not fully satisfied with it.